### ISSUE TITLE
Creating a terminal in a basic (Hello World) Theia app hangs the app

### ISSUE THREAD

**Issue by jcortell68** [eclipse-theia/theia#12733]
### Bug Description:
A basic Theia app hangs when you try to create a terminal.

This is reproducible with a Hello World app produced by the Yeoman generator, but because the generator is currently broken (various outstanding issues), use the equivalent working monorepo I have set up:

```
git clone -b master https://github.com/jcortell68/theiasandbox
cd theiasandbox
yarn
yarn start:electron
```

Create a terminal via `Terminal > New Terminal`
Note that a terminal seems to appear (without a prompt) but the Theia app/UI is completely hung at that point.

### Additional Information

- Operating System: Windows
- Theia Version: 1.39

**Comment by vince-fugnitto** [comment]
It seems to work fine both when I generate a new `hello-world` app, or yours (I am on Linux) however.
Could it be caused by https://github.com/eclipse-theia/theia/issues/12734 instead?

**Comment by jcortell68** [comment]
12734 is a separate issue that I opened; I very much doubt they are related.

Let me try on Linux. This could be a Windows-only issue.

**Comment by jcortell68** [comment]
Yeah. Works fine for me on Linux. We'll need someone else to try on Windows and see if it's a "me" issue or a general one.

**Comment by vladarama** [comment]
@jcortell68 
I just tried on Windows and everything seems to work fine. 
Demo:
![terminal hanging](https://github.com/eclipse-theia/theia/assets/86936229/968971a8-30ad-4e6a-a662-2f359e173ebc)

**Comment by jcortell68** [comment]
@vince-fugnitto BTW, I retried 12734 and that one is still reproducible. So confirmed that issue and this one are unrelated.

**Comment by jcortell68** [comment]
Uh oh. The plot thickens. The hang suddenly and unexpectedly started happening for me, and without being able to workaround it by closing the workspace. After a lot of head-scratching, I realized what was going on. The problem reproduced for me only when I launch the app via the VS Code's launch configuration (`Start Electron Backend`). If I launch the app at the terminal using `yarn start:electron`, then the problem doesn't reproduce.

@vladarama Can you try using the launch configuration? Curious to see if you reproduce the issue that way.

**Comment by vladarama** [comment]
Yeah, I can confirm the bug is present on Windows when launching electron from VS Code's launch config.
Demo:
![terminal hanging](https://github.com/eclipse-theia/theia/assets/86936229/0c9dcd65-9943-4d9c-8c11-78ef487c5006)

**Comment by jcortell68** [comment]
@vladarama Quick side question: what tool do you use to create your GIF screen recordings?

**Comment by vladarama** [comment]
@jcortell68 
I use LICEcap for the GIF recordings:
https://www.cockos.com/licecap/

**Comment by vince-fugnitto** [comment]
@jcortell68 I believe the issue is pretty specific to the [theia-extension-generator](https://github.com/eclipse-theia/generator-theia-extension) (might even be fixed by https://github.com/eclipse-theia/generator-theia-extension/pull/173).

Also note that the `.vscode/launch.json` you have in your repository does not work (@vladarama had to modify https://github.com/jcortell68/theiasandbox/commit/6ed8dd5a7ff3b54f8e3a7e1a7c956cd45868f3cf), and the steps you highlighted to reproduce are incomplete (`yarn` is not sufficient before starting to debug with `start without debugging`).

**Comment by jcortell68** [comment]
@vince-fugnitto Yeah. I just realized the issues with the setup a little while ago. I tried building from a fresh git clone and ran into them. I'm sorting that out now. But I don't _think_ the issue is fixed by the pull request you mention, as I have that change in other branches in my repo and I think I can reproduce the problem there. Anyway, let me clean this all up and I'll post and update here shortly

**Comment by jcortell68** [comment]
OK. I've push a new commit which resolves all the setup issues. I also added a missing step in the description that only made itself obvious in a clean environment (I added `yarn rebuild:electron`)

I git clone'd to a new location and test the reproducibility steps. Everything went smooth, and the problem still reproduces.

I can also confirm it also reproduced only on Windows.

**Comment by Gerd-Augsburg** [comment]
I have a very different environment, but the exact same Problem.
I build the Docker Container from: https://github.com/eclipse-theia/theia-blueprint 
If i try to open a terminal in the browser, the backend hangs.
No Error, just not responsive anymore.

If i restart the backend and open with a browser that "remembers" the open terminal tab i see a warning in the Logs:
"terminal WARN Couldn't attach - can't find terminal with id: -1"
and the backend immediately hangs again. 

Not sure if realated or complete different Problem.

**Comment by jcortell68** [comment]
@Gerd-Augsburg That second scenario also happens for me in the Electron case. I.e., if when I launch the app, it already has a terminal open, then the app hangs right away. I suspect we are running into the same underlying problem.

**Comment by Gerd-Augsburg** [comment]
An update:
The underlying problem might be the same (i assume it can't open a new tty) but the reason for it on my side is Docker.
Local without container it works fine and running the container as a pod in a kubernetes cluster works too.
Only Docker run doesn't work.
If someone knows the docker setting i am missing, it would be nice to know, but here is not the place for docker questions.
So for this thread my problem is not existent any more.

**Comment by martin-fleck-at** [comment]
So I debugged into this on Windows and I believe we are a victim of this bug: https://github.com/microsoft/node-pty/issues/532. If during debugging I pose before line
```
const connect = (this._ptyNative as IConptyNative).connect(this._pty, commandLine, cwd, env, c => this._$onProcessExit(c));
```
then the process runs through. It seems that there is something not ready yet to connect but I couldn't find any stack trace or error.

I tried several versions to see if the problem might be resolved in any of them:
- Latest [1.1.0-beta4](https://www.npmjs.com/package/node-pty/v/1.1.0-beta4) version
- Release [1.0.0](https://www.npmjs.com/package/node-pty/v/1.0.0) 
- Version [0.10.0](https://www.npmjs.com/package/node-pty/v/0.10.0) mentioned in bug report that is supposed to work

Unfortunately, for me they all have the same issue so I think we may need to wait for this bug to be fixed on the node-pty side or we could try to artifically introduce some delay somehow.

**Comment by jcortell68** [comment]
Nice sleuthing. It's unfortunate that node-pty ticket was opened 18 months ago and has had no meaningful activity. I guess waiting it out is reasonable given that it's a bit of a specialize scenario (hang happens on Windows only, only if launched via VS Code launch config, only when launching a terminal). I guess if we see more people pile on to this ticket, we can consider an ugly sleep hack. Thanks for investigating this!

**Comment by akhileshraju** [comment]
Hey all, I see the same issue in a custom application we are building on top of Theia. From what I've read in this issue, it seems this is an upstream issue in `node-pty`. However, as of Jan, [seems it wasn't addressed](https://github.com/eclipse-theia/theia/pull/14677#pullrequestreview-2532168181).
@jcortell68 you mention "an ugly sleep hack". Could you share details? I am interested in using it in my sandbox since the fix doesn't seem to be forthcoming in `node-pty`.

FWIW, I seem the same issue with both the browser launch config and the electron launch config in VS Code. Is it a fair statement to say this is expected since the node-pty bug is not an electron only thing?

**Comment by msujew** [comment]
I believe this issue might have been fixed in the latest stable version of node-pty (1.1.0). I'm currently in the process of upgrading Theia to it, but it might take a few more days. Bundling the dependency is suddenly much more complicated - I'm thinking about declaring the native packages as external instead.

**Comment by msujew** [comment]
It seems like the most recent release still doesn't fix the issue. However just yesterday https://github.com/microsoft/node-pty/pull/885 has been merged, which should fix this. I'll come back to this once a beta release of the patch is available.

**Comment by msujew** [comment]
This has been fixed by https://github.com/microsoft/node-pty/pull/885 and the respective update PR on our end https://github.com/eclipse-theia/theia/pull/17207.
